File: ios/locus/Sources/locus/Core/EngineCallbackBroker.swift
import Foundation

protocol CallbackDeadlineCancellation: AnyObject {
    func cancel()
}

typealias CallbackDeadlineScheduler = (
    _ delay: TimeInterval,
    _ callback: @escaping () -> Void
) -> CallbackDeadlineCancellation

private final class DispatchWorkItemDeadline: CallbackDeadlineCancellation {
    private let workItem: DispatchWorkItem

    init(_ workItem: DispatchWorkItem) {
        self.workItem = workItem
    }

    func cancel() {
        workItem.cancel()
    }
}

func mainQueueCallbackDeadlineScheduler(
    delay: TimeInterval,
    callback: @escaping () -> Void
) -> CallbackDeadlineCancellation {
    let workItem = DispatchWorkItem(block: callback)
    DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: workItem)
    return DispatchWorkItemDeadline(workItem)
}

/// Gives engine-scoped Dart callbacks process-owned completion semantics.
///
/// Each request is tied to an engine owner/generation, bounded by a deadline,
/// retried once on the newest different live engine, and then completed through
/// the existing headless/default fallback. Superseded and late replies are
/// ignored so the native caller observes exactly one completion.
final class EngineCallbackBroker {
    private let bindings: EngineBindingRegistry
    private let deadlineScheduler: CallbackDeadlineScheduler
    private let timeout: TimeInterval
    private let lock = NSLock()
    private var pending: [UInt64: PendingEngineCallback] = [:]
    private var nextRequestID: UInt64 = 0

    init(
        bindings: EngineBindingRegistry,
        deadlineScheduler: @escaping CallbackDeadlineScheduler,
        timeout: TimeInterval = 10
    ) {
        self.bindings = bindings
        self.deadlineScheduler = deadlineScheduler
        self.timeout = timeout
    }

    func request<Result>(
        invoke: @escaping (AnyObject, @escaping (Result) -> Void) -> Void,
        fallback: @escaping (@escaping (Result) -> Void) -> Void,
        terminalDefault: @escaping () -> Result,
        completion: @escaping (Result) -> Void
    ) {
        let requestID: UInt64 = withLock {
            nextRequestID += 1
            return nextRequestID
        }
        let operation = EngineCallbackOperation(
            bindingProvider: { [bindings] excluded in
                bindings.activeSnapshot(excludingOwners: excluded)
            },
            bindingIsLive: { [bindings] snapshot in
                bindings.contains(snapshot)
            },
            deadlineScheduler: deadlineScheduler,
            timeout: timeout,
            invoke: invoke,
            fallback: fallback,
            terminalDefault: terminalDefault,
            completion: completion,
            onFinish: { [weak self] in
                guard let self else { return }
                self.withLock { self.pending.removeValue(forKey: requestID) }
            }
        )
        withLock { pending[requestID] = operation }
        operation.start()
    }

    func ownerDetached(_ owner: AnyObject) {
        let affected = withLock { pending.values.filter { $0.isOwned(by: owner) } }
        affected.forEach { $0.ownerDetached(owner) }
    }

    func bindingClaimed(_ binding: EngineBindingSnapshot) {
        let affected = withLock {
            pending.values.filter {
                $0.isOlderGeneration(owner: binding.owner, generation: binding.generation)
            }
        }
        affected.forEach {
            $0.bindingReplaced(owner: binding.owner, generation: binding.generation)
        }
    }

    @discardableResult
    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}

private protocol PendingEngineCallback: AnyObject {
    func isOwned(by owner: AnyObject) -> Bool
    func isOlderGeneration(owner: AnyObject, generation: UInt64) -> Bool
    func ownerDetached(_ owner: AnyObject)
    func bindingReplaced(owner: AnyObject, generation: UInt64)
}

private final class EngineCallbackOperation<Result>: PendingEngineCallback {
    private static var maxEngineAttempts: Int { 2 }

    private let bindingProvider: ([AnyObject]) -> EngineBindingSnapshot?
    private let bindingIsLive: (EngineBindingSnapshot) -> Bool
    private let deadlineScheduler: CallbackDeadlineScheduler
    private let timeout: TimeInterval
    private let invoke: (AnyObject, @escaping (Result) -> Void) -> Void
    private let fallback: (@escaping (Result) -> Void) -> Void
    private let terminalDefault: () -> Result
    private let completion: (Result) -> Void
    private let onFinish: () -> Void
    private let lock = NSLock()
    private var attemptedOwners: [AnyObject] = []
    private var activeBinding: EngineBindingSnapshot?
    private var activeToken: UInt64 = 0
    private var deadline: CallbackDeadlineCancellation?
    private var completed = false
    private var fallbackStarted = false

    init(
        bindingProvider: @escaping ([AnyObject]) -> EngineBindingSnapshot?,
        bindingIsLive: @escaping (EngineBindingSnapshot) -> Bool,
        deadlineScheduler: @escaping CallbackDeadlineScheduler,
        timeout: TimeInterval,
        invoke: @escaping (AnyObject, @escaping (Result) -> Void) -> Void,
        fallback: @escaping (@escaping (Result) -> Void) -> Void,
        terminalDefault: @escaping () -> Result,
        completion: @escaping (Result) -> Void,
        onFinish: @escaping () -> Void
    ) {
        self.bindingProvider = bindingProvider
        self.bindingIsLive = bindingIsLive
        self.deadlineScheduler = deadlineScheduler
        self.timeout = timeout
        self.invoke = invoke
        self.fallback = fallback
        self.terminalDefault = terminalDefault
        self.completion = completion
        self.onFinish = onFinish
    }

    func start() {
        advanceFromEngine(expectedToken: nil)
    }

    func isOwned(by owner: AnyObject) -> Bool {
        withLock { !completed && activeBinding?.owner === owner }
    }

    func isOlderGeneration(owner: AnyObject, generation: UInt64) -> Bool {
        withLock {
            !completed &&
                activeBinding?.owner === owner &&
                (activeBinding?.generation ?? generation) < generation
        }
    }

    func ownerDetached(_ owner: AnyObject) {
        let token: UInt64? = withLock {
            activeBinding?.owner === owner ? activeToken : nil
        }
        guard let token else { return }
        advanceFromEngine(expectedToken: token)
    }

    func bindingReplaced(owner: AnyObject, generation: UInt64) {
        let token: UInt64? = withLock {
            guard activeBinding?.owner === owner,
                  let currentGeneration = activeBinding?.generation,
                  currentGeneration < generation else {
                return nil
            }
            return activeToken
        }
        guard let token else { return }
        advanceFromEngine(expectedToken: token)
    }

    private func advanceFromEngine(expectedToken: UInt64?) {
        let transition: (token: UInt64, exclusions: [AnyObject], deadline: CallbackDeadlineCancellation?)? = withLock {
            guard !completed, !fallbackStarted else { return nil }
            if let expectedToken, expectedToken != activeToken { return nil }
            let oldDeadline = deadline
            deadline = nil
            activeBinding = nil
            activeToken += 1
            return (activeToken, attemptedOwners, oldDeadline)
        }
        guard let transition else { return }
        transition.deadline?.cancel()

        let next = transition.exclusions.count < Self.maxEngineAttempts
            ? bindingProvider(transition.exclusions)
            : nil
        if let next {
            beginEngineAttempt(next, transitionToken: transition.token)
        } else {
            beginFallback(transitionToken: transition.token)
        }
    }

    private func beginEngineAttempt(
        _ binding: EngineBindingSnapshot,
        transitionToken: UInt64
    ) {
        let shouldInvoke = withLock {
            guard !completed,
                  !fallbackStarted,
                  activeToken == transitionToken else {
                return false
            }
            attemptedOwners.append(binding.owner)
            activeBinding = binding
            return true
        }
        guard shouldInvoke else { return }
        guard bindingIsLive(binding) else {
            advanceFromEngine(expectedToken: transitionToken)
            return
        }

        installDeadline(token: transitionToken) { [weak self] in
            self?.advanceFromEngine(expectedToken: transitionToken)
        }
        let isCurrent = withLock {
            !completed &&
                activeToken == transitionToken &&
                activeBinding?.owner === binding.owner &&
                activeBinding?.generation == binding.generation
        }
        guard isCurrent else { return }
        invoke(binding.owner) { [weak self] value in
            self?.completeIfCurrent(token: transitionToken, value: value)
        }
    }

    private func beginFallback(transitionToken: UInt64) {
        let shouldInvoke = withLock {
            guard !completed,
                  !fallbackStarted,
                  activeToken == transitionToken else {
                return false
            }
            fallbackStarted = true
            return true
        }
        guard shouldInvoke else { return }

        installDeadline(token: transitionToken) { [weak self] in
            guard let self else { return }
            self.completeIfCurrent(
                token: transitionToken,
                value: self.terminalDefault()
            )
        }
        fallback { [weak self] value in
            self?.completeIfCurrent(token: transitionToken, value: value)
        }
    }

    private func installDeadline(token: UInt64, callback: @escaping () -> Void) {
        let cancellation = deadlineScheduler(timeout, callback)
        let shouldCancel = withLock {
            if completed || token != activeToken {
                return true
            }
            deadline = cancellation
            return false
        }
        if shouldCancel { cancellation.cancel() }
    }

    private func completeIfCurrent(token: UInt64, value: Result) {
        let transition: (accepted: Bool, cancellation: CallbackDeadlineCancellation?) = withLock {
            guard !completed, token == activeToken else { return (false, nil) }
            completed = true
            let currentDeadline = deadline
            deadline = nil
            activeBinding = nil
            return (true, currentDeadline)
        }
        guard transition.accepted else { return }
        transition.cancellation?.cancel()
        onFinish()
        completion(value)
    }

    @discardableResult
    private func withLock<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}

File: android/src/main/kotlin/dev/locus/core/EngineCallbackBroker.kt
package dev.locus.core

internal fun interface CallbackDeadlineCancellation {
    fun cancel()
}

internal fun interface CallbackDeadlineScheduler {
    fun schedule(delayMs: Long, callback: () -> Unit): CallbackDeadlineCancellation
}

/**
 * Gives engine-scoped Dart callbacks process-owned completion semantics.
 *
 * A Flutter engine may detach after native code selects its bridge but before
 * Dart replies. Flutter drops replies for a dead messenger, so every request is
 * bounded, tied to the selected owner/generation, retried once on a different
 * live engine, and then completed through the caller's existing fallback.
 */
internal class EngineCallbackBroker<T : Any>(
    private val bindings: EngineBindingRegistry<T>,
    private val deadlineScheduler: CallbackDeadlineScheduler,
    private val timeoutMs: Long = DEFAULT_TIMEOUT_MS,
) {
    private val lock = Any()
    private val pending = mutableMapOf<Long, PendingOperation>()
    private var nextRequestId = 0L

    fun <R> request(
        invoke: (T, (R) -> Unit) -> Unit,
        fallback: ((R) -> Unit) -> Unit,
        terminalDefault: () -> R,
        completion: (R) -> Unit,
    ) {
        val operation = synchronized(lock) {
            nextRequestId += 1
            Operation(
                requestId = nextRequestId,
                invoke = invoke,
                fallback = fallback,
                terminalDefault = terminalDefault,
                completion = completion,
            ).also { pending[nextRequestId] = it }
        }
        operation.start()
    }

    fun ownerDetached(owner: Any) {
        val affected = synchronized(lock) {
            pending.values.filter { it.isOwnedBy(owner) }
        }
        affected.forEach { it.ownerDetached(owner) }
    }

    /**
     * Invalidates an attempt that captured an older bridge generation for the
     * same owner. The owner remains excluded from retry: a replacement on the
     * same engine is not proof that its messenger is a different live route.
     */
    fun bindingClaimed(binding: EngineBindingSnapshot<T>) {
        val affected = synchronized(lock) {
            pending.values.filter {
                it.isOlderGeneration(binding.owner, binding.generation)
            }
        }
        affected.forEach {
            it.bindingReplaced(binding.owner, binding.generation)
        }
    }

    private interface PendingOperation {
        fun isOwnedBy(owner: Any): Boolean
        fun isOlderGeneration(owner: Any, generation: Long): Boolean
        fun ownerDetached(owner: Any)
        fun bindingReplaced(owner: Any, generation: Long)
    }

    private inner class Operation<R>(
        private val requestId: Long,
        private val invoke: (T, (R) -> Unit) -> Unit,
        private val fallback: ((R) -> Unit) -> Unit,
        private val terminalDefault: () -> R,
        private val completion: (R) -> Unit,
    ) : PendingOperation {
        private val attemptedOwners = mutableListOf<Any>()
        private var activeBinding: EngineBindingSnapshot<T>? = null
        private var activeToken = 0L
        private var deadline: CallbackDeadlineCancellation? = null
        private var completed = false
        private var fallbackStarted = false

        fun start() {
            advanceFromEngine(expectedToken = null)
        }

        override fun isOwnedBy(owner: Any): Boolean = synchronized(lock) {
            !completed && activeBinding?.owner === owner
        }

        override fun isOlderGeneration(owner: Any, generation: Long): Boolean = synchronized(lock) {
            !completed &&
                activeBinding?.let {
                    it.owner === owner && it.generation < generation
                } == true
        }

        override fun ownerDetached(owner: Any) {
            val token = synchronized(lock) {
                activeBinding?.takeIf { it.owner === owner }?.let { activeToken }
            } ?: return
            advanceFromEngine(token)
        }

        override fun bindingReplaced(owner: Any, generation: Long) {
            val token = synchronized(lock) {
                activeBinding?.takeIf {
                    it.owner === owner && it.generation < generation
                }?.let { activeToken }
            } ?: return
            advanceFromEngine(token)
        }

        private fun advanceFromEngine(expectedToken: Long?) {
            val transition = synchronized(lock) {
                if (completed || fallbackStarted) return
                if (expectedToken != null && expectedToken != activeToken) return
                val previousDeadline = deadline
                deadline = null
                activeBinding = null
                activeToken += 1
                Transition(
                    token = activeToken,
                    exclusions = attemptedOwners.toList(),
                    previousDeadline = previousDeadline,
                )
            }
            transition.previousDeadline?.cancel()

            val next = if (transition.exclusions.size < MAX_ENGINE_ATTEMPTS) {
                bindings.currentSnapshot(excludingOwners = transition.exclusions)
            } else {
                null
            }
            if (next == null) {
                beginFallback(transition.token)
            } else {
                beginEngineAttempt(next, transition.token)
            }
        }

        private fun beginEngineAttempt(
            binding: EngineBindingSnapshot<T>,
            transitionToken: Long,
        ) {
            val shouldInvoke = synchronized(lock) {
                if (completed || fallbackStarted || transitionToken != activeToken) {
                    return@synchronized false
                }
                attemptedOwners += binding.owner
                activeBinding = binding
                true
            }
            if (!shouldInvoke) return
            if (!bindings.contains(binding)) {
                advanceFromEngine(transitionToken)
                return
            }
            installDeadline(transitionToken) { advanceFromEngine(transitionToken) }
            val isCurrent = synchronized(lock) {
                !completed &&
                    transitionToken == activeToken &&
                    activeBinding == binding
            }
            if (!isCurrent) return
            try {
                invoke(binding.value) { value -> completeIfCurrent(transitionToken, value) }
            } catch (_: RuntimeException) {
                advanceFromEngine(transitionToken)
            }
        }

        private fun beginFallback(transitionToken: Long) {
            val shouldInvoke = synchronized(lock) {
                if (completed || fallbackStarted || transitionToken != activeToken) {
                    return@synchronized false
                }
                fallbackStarted = true
                true
            }
            if (!shouldInvoke) return
            installDeadline(transitionToken) {
                completeIfCurrent(transitionToken, terminalDefault())
            }
            try {
                fallback { value -> completeIfCurrent(transitionToken, value) }
            } catch (_: RuntimeException) {
                completeIfCurrent(transitionToken, terminalDefault())
            }
        }

        private fun installDeadline(token: Long, onDeadline: () -> Unit) {
            val cancellation = deadlineScheduler.schedule(timeoutMs, onDeadline)
            synchronized(lock) {
                if (completed || token != activeToken) {
                    cancellation.cancel()
                } else {
                    deadline = cancellation
                }
            }
        }

        private fun completeIfCurrent(token: Long, value: R) {
            val callback = synchronized(lock) {
                if (completed || token != activeToken) return
                completed = true
                deadline?.cancel()
                deadline = null
                activeBinding = null
                pending.remove(requestId)
                completion
            }
            callback(value)
        }

    }

    private data class Transition(
        val token: Long,
        val exclusions: List<Any>,
        val previousDeadline: CallbackDeadlineCancellation?,
    )

    private companion object {
        const val DEFAULT_TIMEOUT_MS = 10_000L
        const val MAX_ENGINE_ATTEMPTS = 2
    }
}
